File: nim_m/lexer.nim
import strutils
import token

type
  Lexer* = ref object
    input: string
    position: int     # current position in input (points to current char)
    nextPosition: int # current reading position in input (after current char)
    ch: char          # current char under examination
    size: int         # 変更箇所

proc isLetter(ch: char): bool =
  ('a' <= ch and ch <= 'z') or ('A' <= ch and ch <= 'Z') or (ch == '_')

proc isDigit(ch: char): bool =
  ('0' <= ch and ch <= '9') or (ch == '.')

proc readChar(lex: Lexer) =
  if lex.nextPosition >= lex.size:
    lex.ch = '\0'
  else:
    lex.ch = lex.input[lex.nextPosition]
  lex.position = lex.nextPosition
  lex.nextPosition += 1

proc peekChar(lex: Lexer): char =
  if lex.nextPosition >= lex.size:
    return '\0'
  lex.input[lex.nextPosition]

proc skipWhitespace(lex: Lexer) =
  while lex.ch == ' ' or lex.ch == '\t' or lex.ch == '\n' or lex.ch == '\r':
    lex.readChar()

proc readIdentifier(lex: Lexer): string =
  let position = lex.position
  while isLetter(lex.ch):
    lex.readChar()
  lex.input[position..<lex.position]

proc readNumber(lex: Lexer): string =
  let position = lex.position
  while isDigit(lex.ch):
    lex.readChar()
  lex.input[position..<lex.position]

proc readString(lex: Lexer): string =
  let position = lex.position + 1
  while true:
    lex.readChar()
    if lex.ch == '"' or lex.ch == '\0':
      break
  return lex.input[position..<lex.position]

proc LexerNew*(input: string): Lexer =
  result = new Lexer
  result.input = input
  result.size = len(input)
  result.readChar()

proc nextToken*(lex: Lexer): Token =
  result = new Token

  lex.skipWhitespace()

  case lex.ch
  of '=':
    if lex.peekChar() == '=':
      lex.readChar()
      result.tokenType = EQ
      result.literal = "=="
    else:
      result.tokenType = ASSIGN
      # '$'を使用して、文字を単一の文字列に変換できます
      result.literal = $lex.ch
  of '+':
    result.tokenType = PLUS
    result.literal = $lex.ch
  of '-':
    result.tokenType = MINUS
    result.literal = $lex.ch
  of '!':
    if lex.peekChar() == '=':
      lex.readChar()
      result.tokenType = NOT_EQ
      result.literal = "!="
    else:
      result.tokenType = BANG
      result.literal = $lex.ch
  of '/':
    result.tokenType = SLASH
    result.literal = $lex.ch
  of '*':
    result.tokenType = ASTERISK
    result.literal = $lex.ch
  of '<':
    result.tokenType = LT
    result.literal = $lex.ch
  of '>':
    result.tokenType = GT
    result.literal = $lex.ch
  of ';':
    result.tokenType = SEMICOLON
    result.literal = $lex.ch
  of ',':
    result.tokenType = COMMA
    result.literal = $lex.ch
  of '{':
    result.tokenType = LBRACE
    result.literal = $lex.ch
  of '}':
    result.tokenType = RBRACE
    result.literal = $lex.ch
  of '(':
    result.tokenType = LPAREN
    result.literal = $lex.ch
  of ')':
    result.tokenType = RPAREN
    result.literal = $lex.ch
  of '"':
    result.tokenType = STRING
    result.literal = lex.readString()
  of '\0':
    result.tokenType = EOF
    result.literal = ""
  else:
    if isLetter(lex.ch):
      result.literal = lex.readIdentifier()
      result.tokenType = lookupIdent(result.literal)
      return result
    elif isDigit(lex.ch):
      let literal = lex.readNumber()
      if literal.count(".") == 0:
        result.tokenType = INT
        result.literal = literal
        return result
      elif literal.count(".") == 1:
        result.tokenType = FLOAT
        result.literal = literal
        return result
      else:
        result.tokenType = ILLEGAL
        result.literal = literal
    else:
      result.tokenType = ILLEGAL
      result.literal = $lex.ch

  lex.readChar()

File: py_m/lexer_.py
from token_ import TokenType, Token


class Lexer:
    """字句解析"""

    def __init__(self, input, position=0, next_position=0, ch=""):
        self.input = input
        self.position = position
        self.next_position = next_position
        self.ch = ch
        self.size = len(self.input)

        self.read_char()

    def read_char(self):
        if self.next_position >= self.size:
            self.ch = ""
        else:
            self.ch = self.input[self.next_position]

        self.position = self.next_position
        self.next_position += 1

    def skip_whitespace(self):
        while self.ch == " " or self.ch == "\t" or self.ch == "\n" or self.ch == "\r":
            self.read_char()

    @staticmethod
    def is_letter(v):
        if v.isalpha():
            return True
        elif v == "_":
            return True
        else:
            return False

    @staticmethod
    def is_digit(v):
        if v.isdigit():
            return True
        elif v == '.':
            return True
        else:
            return False

    def peek_char(self):
        if self.next_position >= self.size:
            return ""
        else:
            return self.input[self.next_position]

    def read_identifier(self):
        position = self.position
        while self.is_letter(self.ch):
            self.read_char()
        return self.input[position:self.position]

    def read_number(self):
        position = self.position
        while self.is_digit(self.ch):
            self.read_char()

        return self.input[position:self.position]

    def read_string(self):
        position = self.position + 1
        while True:
            self.read_char()
            if self.ch == '"' or self.ch == "":
                break
        return self.input[position:self.position]

    def next_token(self):
        tok = Token()

        self.skip_whitespace()

        if self.ch == "=":
            if self.peek_char() == "=":
                self.read_char()
                tok.token_type = TokenType.EQ
                tok.literal = "=="
            else:
                tok.token_type = TokenType.ASSIGN
                tok.literal = "="
        elif self.ch == "+":
            tok.token_type = TokenType.PLUS
            tok.literal = self.ch
        elif self.ch == "-":
            tok.token_type = TokenType.MINUS
            tok.literal = self.ch
        elif self.ch == "!":
            if self.peek_char() == "=":
                self.read_char()
                tok.token_type = TokenType.NOT_EQ
                tok.literal = "!="
            else:
                tok.token_type = TokenType.BANG
                tok.literal = "!"
        elif self.ch == "/":
            tok.token_type = TokenType.SLASH
            tok.literal = self.ch
        elif self.ch == "*":
            tok.token_type = TokenType.ASTERISK
            tok.literal = self.ch
        elif self.ch == "<":
            tok.token_type = TokenType.LT
            tok.literal = self.ch
        elif self.ch == ">":
            tok.token_type = TokenType.GT
            tok.literal = self.ch
        elif self.ch == ";":
            tok.token_type = TokenType.SEMICOLON
            tok.literal = self.ch
        elif self.ch == ",":
            tok.token_type = TokenType.COMMA
            tok.literal = self.ch
        elif self.ch == "{":
            tok.token_type = TokenType.LBRACE
            tok.literal = self.ch
        elif self.ch == "}":
            tok.token_type = TokenType.RBRACE
            tok.literal = self.ch
        elif self.ch == "(":
            tok.token_type = TokenType.LPAREN
            tok.literal = self.ch
        elif self.ch == ")":
            tok.token_type = TokenType.RPAREN
            tok.literal = self.ch
        elif self.ch == '"':
            tok.token_type = TokenType.STRING
            tok.literal = self.read_string()
        elif self.ch == "[":
            tok.token_type = TokenType.LBRACKET
            tok.literal = self.ch
        elif self.ch == "]":
            tok.token_type = TokenType.RBRACKET
            tok.literal = self.ch
        elif self.ch == "":
            tok.token_type = TokenType.EOF
            tok.literal = ""
        else:
            if self.is_letter(self.ch):
                tok.literal = self.read_identifier()
                tok.token_type = tok.lookup_ident(tok.literal)
                return tok
            elif self.is_digit(self.ch):
                literal = self.read_number()
                if literal.count(".") == 0:
                    tok.token_type = TokenType.INT
                    tok.literal = literal
                    return tok
                elif literal.count(".") == 1:
                    tok.token_type = TokenType.FLOAT
                    tok.literal = literal
                    return tok
                else:
                    tok.token_type = TokenType.ILLEGAL
                    tok.literal = literal
            else:
                tok.token_type = TokenType.ILLEGAL
                tok.literal = self.ch

        self.read_char()
        return tok

    def __str__(self):
        return "Lexer()"


if __name__ == "__main__":
    pass
